### ISSUE TITLE
Install on Raspberry Pi

### ISSUE THREAD

**Issue by Becker884** [halfgaar/FlashMQ#85]
Can you create a tutorial how to install FlasMQ on RaspberryPi ?
I tried but failed:
`N: Skipping loading the configured file "main/binary-armhf/Packages" because the depot "http://repo.flashmq.org/apt bullseye InRelease" **does not support the "armhf**" architecture.`

If I didn't know for sure that it would run on another RaspberryPi (Victron Venus) - I would not ask for.

**Comment by halfgaar** [comment]
Adding Debian (based) ARM packages is actually a good idea. I would require some time to set that up though.

In the mean time, building on the Raspberry Pi itself should be easy. You can just follow the instructions in the readme, using `build.sh`, or `cmake` manually. Installing `cmake` and `build-essential` should be enough.

This won't give you a systemd unit file and such, but perhaps it's a start?

**Comment by Becker884** [comment]
Thank you for your quick response.
However, that is beyond my skills ;-) so I´ll try it on my VM with amd64.
I´m using mosquitto and it doesn´t support IPv6 over websockets.
that's why I wanted to try FlashMQ.
can you tell me if FlashMQ support ws(s) over IPv6?
preferably IPv4 & IPv6 at the same time via wss.
mosquitto can only do mqtt over IPv6.

**Comment by halfgaar** [comment]
Yep, FlashMQ indeed supports websockets and TLS websockets over IPv4 and IPv6.

You know, I just realized. If you do compile on a Rpi, cmake probably does make a .deb package you can just install with `dpkg -i`.

**Comment by halfgaar** [comment]
BTW, I see you tried the Debian bullseye repo. Does that mean your Rpi also runs that? I was going to provide you with a test binary, but then I need to know for sure which one: bullseye or bookworm.

**Comment by Becker884** [comment]
I tried on bullseye

**Comment by halfgaar** [comment]
Can you test [this deb file](https://www.flashmq.org/wp-content/uploads/flashmq_1.9.1-1708187570+bullseye_armhf.deb)? It's for Bullseye, armhf (32 bit).

You can install with:

```
dpkg -i flashmq_1.9.1-1708187570+bullseye_armhf.deb
```

You will then have a systemd service; check with:

```
systemctl status flashmq
```

**Comment by Becker884** [comment]
![grafik](https://github.com/halfgaar/FlashMQ/assets/81651521/a619bb83-0066-4837-8dc4-cc643acb44bf)

installation seems to be successfull, I hope i´ll get it run

**Comment by Becker884** [comment]
It is the same as with mosquitto:
![grafik](https://github.com/halfgaar/FlashMQ/assets/81651521/6dc5cc72-9d01-4bc1-862f-3d36a19324f6)

`Creating IPv6 non-SSL websocket listener on [::]:4443`

but I can only connect by mqtt IPv6, ws only IPv4.

here: https://github.com/eclipse/mosquitto/issues/2628 ralight wrote 

> I believe it depends on how libwebsockets is compiled. It does not have support for IPv6 compiled in by default.

is this the same problem with FlashMQ?

**Comment by halfgaar** [comment]
> is this the same problem with FlashMQ (libwebsockets) ?

No, websocket frame parsing is implemented internally by FlashMQ.

Can you show logs of the server and your clients?

**Comment by Becker884** [comment]
config:
`# https://www.flashmq.org/documentation/config-file/

log_file    /var/log/flashmq/flashmq.log
storage_dir /var/lib/flashmq
allow_anonymous true

listen {
  protocol mqtt
  port 1883
  inet_protocol ip4_ip6
}

listen {
  protocol websockets
  port 4443
  inet_protocol ip4_ip6
}`

`pi@raspberrypi2:~ $ netstat -tunlp | grep LISTEN
(Es konnten nicht alle Prozesse identifiziert werden; Informationen über
nicht-eigene Processe werden nicht angezeigt; Root kann sie anzeigen.)
tcp        0      0 0.0.0.0:4443            0.0.0.0:*               LISTEN      -
tcp        0      0 0.0.0.0:1883            0.0.0.0:*               LISTEN      -
tcp        0      0 0.0.0.0:22              0.0.0.0:*               LISTEN      -
tcp6       0      0 :::1883                 :::*                    LISTEN      -
tcp6       0      0 :::4443                 :::*                    LISTEN      -
tcp6       0      0 :::22                   :::*                    LISTEN      -`

Log:
`[2024-02-17 21:22:03.133] [NOTICE] Quitting FlashMQ
[2024-02-17 21:22:03.137] [NOTICE] Thread 1 doing auth cleanup.
[2024-02-17 21:22:03.137] [NOTICE] Thread 3 doing auth cleanup.
[2024-02-17 21:22:03.138] [NOTICE] Thread 2 doing auth cleanup.
[2024-02-17 21:22:03.138] [NOTICE] Thread 0 doing auth cleanup.
[2024-02-17 21:22:03.140] [INFO] Saving retained messages to '/var/lib/flashmq/retained.db'
[2024-02-17 21:22:03.161] [INFO] Saving sessions and subscriptions to '/var/lib/flashmq/sessions.db' in thread.
[2024-02-17 21:22:03.165] [INFO] Saved 0 sessions and 0 subscriptions to '/var/lib/flashmq/sessions.db' in 3 ms.
[2024-02-17 21:22:03.183] [INFO] Saving bridge info in '/var/lib/flashmq/bridgenames.db'
[2024-02-17 21:22:03.202] [INFO] Saving states done
[2024-02-17 21:22:03.317] [INFO] Setting rlimit nofile to 1000000.
[2024-02-17 21:22:03.317] [NOTICE] 4 CPUs are detected, making as many threads. Use 'thread_count' setting to override.
[2024-02-17 21:22:03.318] [INFO] Loading '/var/lib/flashmq/retained.db'
[2024-02-17 21:22:03.323] [INFO] Loading '/var/lib/flashmq/sessions.db'
[2024-02-17 21:22:03.327] [NOTICE] Starting FlashMQ version 1.9.1, release build without SSE support.
[2024-02-17 21:22:03.328] [INFO] Switching logging from stdout to logfile '/var/log/flashmq/flashmq.log'
[2024-02-17 21:22:03.335] [NOTICE] Thread 0 doing auth init.
[2024-02-17 21:22:03.338] [NOTICE] Thread 2 doing auth init.
[2024-02-17 21:22:03.338] [NOTICE] Thread 1 doing auth init.
[2024-02-17 21:22:03.343] [NOTICE] Thread 3 doing auth init.
[2024-02-17 21:23:23.482] [NOTICE] Accepting connection from: address='fd00:aaaa::5c45:bc8f:e70d:d044', transport='TCP/Non-SSL', fd=20
[2024-02-17 21:23:23.482] [NOTICE] Client '[ClientID='mqtt-explorer-473ae2c1', username='', fd=20, keepalive=60s, transport='TCP/Non-SSL', address='fd00:aaaa::5c45:bc8f:e70d:d044', prot=3.1.1, clean=1]' logged in successfully
[2024-02-17 21:23:26.311] [NOTICE] Client '[ClientID='mqtt-explorer-473ae2c1', username='', fd=20, keepalive=60s, transport='TCP/Non-SSL', address='fd00:aaaa::5c45:bc8f:e70d:d044', prot=3.1.1, clean=1]' cleanly disconnecting
[2024-02-17 21:23:26.311] [NOTICE] Removing client '[ClientID='mqtt-explorer-473ae2c1', username='', fd=20, keepalive=60s, transport='TCP/Non-SSL', address='fd00:aaaa::5c45:bc8f:e70d:d044', prot=3.1.1, clean=1]'. Reason(s): MQTT Disconnect received (code 0)., socket disconnect detected
[2024-02-17 21:23:37.788] [NOTICE] Accepting connection from: address='192.168.168.20', transport='TCP/Non-SSL', fd=20
[2024-02-17 21:23:37.788] [NOTICE] Client '[ClientID='mqtt-explorer-473ae2c1', username='', fd=20, keepalive=60s, transport='TCP/Non-SSL', address='192.168.168.20', prot=3.1.1, clean=1]' logged in successfully
[2024-02-17 21:23:42.459] [NOTICE] Client '[ClientID='mqtt-explorer-473ae2c1', username='', fd=20, keepalive=60s, transport='TCP/Non-SSL', address='192.168.168.20', prot=3.1.1, clean=1]' cleanly disconnecting
[2024-02-17 21:23:42.460] [NOTICE] Removing client '[ClientID='mqtt-explorer-473ae2c1', username='', fd=20, keepalive=60s, transport='TCP/Non-SSL', address='192.168.168.20', prot=3.1.1, clean=1]'. Reason(s): MQTT Disconnect received (code 0)., socket disconnect detected
[2024-02-17 21:23:56.310] [NOTICE] Accepting connection from: address='192.168.168.20', transport='TCP/Websocket/Non-SSL', fd=20
[2024-02-17 21:23:56.319] [NOTICE] Client '[ClientID='mqtt-explorer-473ae2c1', username='', fd=20, keepalive=60s, transport='TCP/Websocket/Non-SSL', address='192.168.168.20', prot=3.1.1, clean=1]' logged in successfully
[2024-02-17 21:24:03.235] [NOTICE] Client '[ClientID='mqtt-explorer-473ae2c1', username='', fd=20, keepalive=60s, transport='TCP/Websocket/Non-SSL', address='192.168.168.20', prot=3.1.1, clean=1]' cleanly disconnecting
[2024-02-17 21:24:03.235] [NOTICE] Removing client '[ClientID='mqtt-explorer-473ae2c1', username='', fd=20, keepalive=60s, transport='TCP/Websocket/Non-SSL', address='192.168.168.20', prot=3.1.1, clean=1]'. Reason(s): MQTT Disconnect received (code 0)., socket disconnect detected`

MQTT Explorer says only "disconnect from server" if I try WS IPv6

**Comment by halfgaar** [comment]
If there is not even a line saying 'accepting connection', the problem is somewhere else. That is the lowest level of IPv6 activity, when it doesn't even know yet whether it's a websocket or not, or TLS or not.

The problem lies elsewhere. And I've seen other weird stuff with MQTT explorer, like it not respecting a custom port number; it just kept using the default. Can you try another client?

**Comment by Becker884** [comment]
I tried with Node-Red, same.
MQTT IPv4 & IPv6 works
WS IPv4 works
WS IPv6 can´t connect
did you try it successfully ?

**Comment by halfgaar** [comment]
I'm having trouble with test.mosquitto.org, but that could be my client. I'm not really well set up with websockets currently.

Can you try `demo.flashmq.org`, port 4443? That's WSS. That one works for me.

**Comment by Becker884** [comment]
It is the same.
wss://[2a01:1b0:7996:418:83:137:146:230]:4443 doesn´t connect.
wss://83.137.146.230:4443 no problem
mqtt://[2a01:1b0:7996:418:83:137:146:230]:1883 no problem

**Comment by halfgaar** [comment]
And if you do?:

```
telnet -6 <address6> <port>
```

That should at least produce a line in the log.

Or:

```
nmap -sT -p <port> -6 <address6>
```

If those don't show up in the log as

```
Accepting connection from: address='[ipv6address]'
```

Then you have network problems.

**Comment by Becker884** [comment]
I´ve no telnet or nmap :-(

my first characters: 
2003:da:b702:
93.213.74.

**Comment by halfgaar** [comment]
I see logs from you like:

```
[2024-02-18 10:49:11.843] [NOTICE] Accepting connection from: address='2003:da:b702:blabla:e8ca', transport='TCP/Websocket/SSL', fd=25
[2024-02-18 10:49:11.897] [NOTICE] Removing client '[ClientID='', username='', fd=25, keepalive=10s, transport='TCP/Websocket/SSL', address='2003:da:b702:blabla:e8ca', prot=none, clean=0]'. Reason(s): Problem accepting SSL socket: error:14094412:SSL routines:ssl3_read_bytes:sslv3 alert bad certificate
```

Did you use [this certificate](https://www.flashmq.org/wp-content/uploads/2021/03/flashmq-ca.crt_.zip), or set 'insecure' mode?

I also saw succefull ones:

```
[2024-02-18 10:50:39.804] [NOTICE] Client '[ClientID='Browser_7dfd1863', username='', fd=26, keepalive=60s, transport='TCP/Websocket/SSL', address='2003:da:b702:blabla:e8ca', prot=3.1.1, clean=1]' logged in successfully
[2024-02-18 10:50:39.825] [SUBSCRIBE] Client '[ClientID='Browser_7dfd1863', username='', fd=26, keepalive=60s, transport='TCP/Websocket/SSL', address='2003:da:b702:blabla:e8ca', prot=3.1.1, clean=1]' subscribed to 'VLT' QoS 0
[2024-02-18 10:51:01.698] [NOTICE] Removing client '[ClientID='Browser_7dfd1863', username='', fd=26, keepalive=60s, transport='TCP/Websocket/SSL', address='2003:da:b702:blabla:e8ca', prot=3.1.1, clean=1]'. Reason(s): Browser navigating away from page, socket disconnect detected
```

I don't see attempts for non-TLS websockets.

**Comment by Becker884** [comment]
I got it now with JS in html (browser):
`<script src="https://cdnjs.cloudflare.com/ajax/libs/paho-mqtt/1.0.1/mqttws31.min.js"></script>  
<script>
        const clientId = 'Browser_' + Math.random().toString(16).substr(2, 8);  
        client = new Paho.MQTT.Client("2a01:1b0:7996:418:83:137:146:230", Number(8080), clientId);  
        client.onConnectionLost = onConnectionLost;
        client.onMessageArrived = onMessageArrived;
        client.connect({
            onSuccess: onConnect,
            useSSL: false
        });
        function onConnect() {
            console.log("Verbindung hergestellt");  
            client.subscribe("topic/path");  
        }
        function onConnectionLost(responseObject) {
            if (responseObject.errorCode !== 0) {
                console.log("Verbindung verloren:" + responseObject.errorMessage);  
            }
        }
        function onMessageArrived(message) {
            console.log("Beschreibung: " + message.destinationName);  
            console.log("Nachricht: " + message.payloadString);  
        }
</script>`

I´ll try port 4443 (SSL true) with the domain name and disabled IPv4 on my computer.

I think the problem is with MQTT Explorer, it can not even connect without SSL / port 8080 on IPv6.

**Comment by Becker884** [comment]
I tried with disabled IPv4 - you should see my succesfully login :-)

**Comment by halfgaar** [comment]
I indeed see a whole bunch of IPv6 websocket traffic from your client ID `Browser_xxx`, with and without TLS.

As for the config file, see [the documentation](https://www.flashmq.org/man/flashmq.conf.5). Most, if not all, of that is convertible to FlashMQ config directives.

Reloading is indeed done with `systemctl reload flashmq.service`. As for permissions to the certificates and how you renew them, I'll leave that to you. You can look at [FlashMQ's Let's Encrypt](https://www.flashmq.org/documentation/lets-encrypt/) example if you want.

I think this ticket can be closed, I added a new one for the Raspberry Pi builds (#86).
